### PR TITLE
Add carto.us sample data

### PR DESCRIPTION
**Pull request by bryevdv**
This Pull request:

* Modernizes the`bokeh.util.sampledata` module with `pathlib` and better types
* Adds a `carto.us` sampledata that has a projected US map, including AK, HI and other territories re-positioned near the continental states. 

With this PR, the following code is possible:
```python
from bokeh.models import GeoJSONDataSource
from bokeh.plotting import figure, show
from bokeh.sampledata import carto

geo_source = GeoJSONDataSource(geojson=carto.us)
p = figure(match_aspect=True, sizing_mode="stretch_width")
p.patches("xs", "ys", fill_color=None, source=geo_source)
show(p)
```
Which generates:

Screen Shot 2022-06-07 at 09 46 39

### Notes/Questions

The files were discovered at https://community.tableau.com/s/news/a0A4T000002NGVhUAO/new-spatial-files I reached out to the author who simply noted that they were derived from USGov Public Domain sources. For simplicity this is all I note in the docstring.

Since the outlying territories are repositioned, I consider this to be a cartogram, hence the name `carto.us`. I envision we can add similar data sets for other world areas under the `carto` module. I am interested if there are any other ideas about module/data names, however. 

The current data value is a (geo)JSON string, since that is all that `GeoJSONDataSource` accepts. This seems pretty inflexible, since almost certainly users will want to mix in other data with the shapes. To do that now a user would have to convert the sampledata string value to geopandas (or whatever), manipulated it, convert it back. 😦  

Some ideas:

* add `geopandas` as an optional dependency for this sample data source (or is there another better package), and return a real data structure?
* expands the kinds of input `GeoJSONDataSource` can accept, e.g. an actual geojson dict or geopandas df instead of just a string

cc @bokeh/dev for thoughts and also @birdsarah in case you have bandwidth to offer advice.

### CODE REVIEW

File: bokeh/sampledata/carto.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2012 - 2022, Anaconda, Inc., and Bokeh Contributors.
+# All rights reserved.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+''' Provide JSON data for co-occurrence of characters in Les Miserables.
```
Comment (tcmetzger) [inline bokeh/sampledata/carto.py:7]: This doesn't look right! 😀

**Comment by bryevdv** [comment]
@mattpap I don't recall anything as to why `geojson` property is a JSON *string*. It was a a long time ago, so I expect it was probably just some early-days limitation. Can you think of any reasons why the property type couldn't become an actual geojson dict instead (with a JSON string adapter for compat).

**Comment by bryevdv** [comment]
OK I have made a simple change to switch `geojson` from `String` to `Dict`. With that change, things I regard as approaching useful (mixing the geo data with outside data) start to become possible:
```
import json

import geopandas as gpd
import numpy as np

from bokeh.models import GeoJSONDataSource
from bokeh.plotting import figure, show
from bokeh.sampledata.carto import united_states
from bokeh.transform import linear_cmap

points = gpd.GeoDataFrame.from_features(united_states)
points["data"] = np.random.random(len(points))

geo_source = GeoJSONDataSource(geojson=points.to_json())
p = figure(match_aspect=True, sizing_mode="stretch_width")
p.patches("xs", "ys", source=geo_source,
          fill_color=linear_cmap('data', 'Viridis256', 0, 1))
show(p)
```
Screen Shot 2022-06-11 at 10 10 26

I think it would also be a reasonable idea to add `GeoJSONDataSource.from_gdf` to encapsulate optional  geopandas support. 

Thoughts here are welcome. cc @bokeh/dev

**Comment by bryevdv** [comment]
I would add: I am also open to the idea of just making the sampledata itself return a GeoDataFrame. We already have sample data sets that rely on optional pandas support. That would also make previewing the sample data in the ref guide more possible.

**Comment by bryevdv** [comment]
@tcmetzger if you have any ideas for a better example than a random colormap per-territory, please let me know. 

We might also consider trying to simplify/reduce the number of points in the data. Currently I would say interactive zoom, etc is at the low end of tolerable. WebGL for (multi-polygons) will really make this sing (cc @ianthomas23 for any comments about what that might take in the future, I think I recall it being a non-trivial lift)

**Comment by ianthomas23** [comment]
@bryevdv WebGL for multi polygons is part of the NumFOCUS SDG that I recently submitted so a solution is on the horizon.

**Comment by ianthomas23** [comment]
Could color by population density as shown in https://en.wikipedia.org/wiki/List_of_states_and_territories_of_the_United_States_by_population_density

**Comment by mattpap** [comment]
> geojson property is a JSON string

`geojson` implies this property accepts JSON and JSON is a string containing a JavaScript-like syntax, which we often represent with related data structures, but technically speaking that's not JSON anymore. I suppose, if someone is going to use this property, then they already have some geojson that they can paste in there. Not a user, so I'm not sure about specific use cases. Perhaps there were other considerations, maybe serialization/protocol related, like handling of NaNs or performance. 

Depending on the size of the input, performance is something to consider when serializing and deserializing a lot of small objects back and forth. It's definitely more efficient to just keep the input as a string, especially if we had a smarter deserializer of GeoJson in bokehjs (streaming, selective, etc.).

One thing to consider is preservation of the original data. "Fortunately" in this case, JSON doesn't support comments or other kinds of metadata, so we can convert losslessly between JSON and equivalent data structures. I'm generally not in favour of bokeh touching users' data in any ways, unless no other options are available. It can and should communicate possible inefficiencies though.

I would allow both types of input, but when used, parse JSON only in bokehjs.

**Comment by tcmetzger** [comment]
@bryevdv I think that @ianthomas23's idea for population density would be very easy to understand as a dataset and would make a great example! Alternatively, we could use the unemployment data that we already have in the sampledata module (grouped by `state_id`): https://docs.bokeh.org/en/2.4.0/docs/reference/sampledata.html#module-bokeh.sampledata.unemployment

**Comment by bryevdv** [comment]
>  It's definitely more efficient to just keep the input as a string, especially if we had a smarter deserializer of GeoJson in bokehjs (streaming, selective, etc.).

I guess you'd have to elaborate on this. If their is just one giant string property, I don't see how a user can update or change anything, without re-sending everything. Unless we add specialized streaming/patching events is certainly possible.

Ideally, what would actually be nice is an ability to load a static geojson source for a map once, and then update columns of associated data entirely separately (rather than properties directly in the geojson itself).  E.g. in an interactive map of  "unemployment in the US" with a slider to move years, the map data is never the part that is changing. However to do that, we would need to implement ways to do the joining between some "feature id" in the geo data and in the columns in BokehJS, since there is no way to rely on the data in the features matching the order of data in the columns. That's definitely more than I can bite off right now. 

However, if string is better as the fundamental type for now, I can revert that change, and instead offer offer `.accepts` and `from_gdf` conveniences to support user needs.

**Comment by mattpap** [comment]
> (...) and instead offer offer .accepts and (...)

I actually don't like `.accepts()` in such cases, and prefer to leave the input as-is. So, I would allow `Either(Dict(...), String)` and deserialize the string case in bokehjs. I prefer to leave `.accepts()` for use cases like in PR #12170.

**Comment by bryevdv** [comment]
@mattpap I will have to try to do some actual timings, since I think that the `String` version may be faster, as you suggested. At least, my initial subjective response after switching to `Dict` was "this seems to be perceptibly slower". Wouldn't another option be keep the user value but add a custom serialization path that uses `json.dumps` directly on the Python side, if the value is a `Dict`? At least, there used to be hooks like that for overriding default serialization. 

Also, I have not gotten any feedback about `geopandas` for the sampledata. Presently I think I am inclined to make it a dependency for this data set.

**Comment by bryevdv** [comment]
Just a heads up I am going to use geopandas in any sampledata files that make sense, e.g the `us_counties`, etc that are currently a awkward and non-standard dict format.

**Comment by bryevdv** [comment]
Having some second thoughts about this as geojson is technically supposed to be lat/lon coordinates only. Would need to recompute the data file and then do a mercator conversion in the example code

**Comment by bryevdv** [comment]
This is quite out of date,  I am going to close this PR but keep the branch for reference for now.

**Pull request closed without merging**